### DISCUSSION TITLE
Generate Typescript classes from self hosted docker supabase instance

### DISCUSSION THREAD

**Discussion by Hatzen** [supabase/supabase#11070] [Questions] [answered]
I tried to generate the typescript schema for my local environment, but i can not get it work.  

from the docs i got quite far but failing to connect to my local database even without changing any password.  
(this commands are not tested but should represent what i have done so far.)
```
git clone https://github.com/supabase/supabase.git
cd supabase
cp docker/.env.example docker/.env
docker-compose -f /docker/docker-compose.yml up -d
npm install supabase --save-dev
npx supabase init
npx supabase start
npx supabase gen types typescript --schema public --db-url postgres://postgres:your-super-secret-and-long-postgres-password@localhost:5432/postgres --debug
```
But then i get an connection refused error. 
```
....
2022/12/17 14:00:18 Recv First Byte
2022/12/17 14:00:18 Sent Header: Host [docker]
2022/12/17 14:00:18 Sent Header: User-Agent [Go-http-client/1.1]
2022/12/17 14:00:18 Send Done
2022/12/17 14:00:18 Recv First Byte
/usr/src/app/bin/src/server/app.js:75
            throw new Error(schemasError.message);
                  ^

Error: connect ECONNREFUSED 127.0.0.1:5432
    at /usr/src/app/bin/src/server/app.js:75:19
    at Generator.next (<anonymous>)
    at fulfilled (/usr/src/app/bin/src/server/app.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
2022/12/17 14:00:18 Sent Header: Host [docker]
2022/12/17 14:00:18 Sent Header: User-Agent [Go-http-client/1.1]
2022/12/17 14:00:18 Send Done
2022/12/17 14:00:19 Recv First Byte
Error: error running container
```  
When changing anything of the db-url i get connection timeout, so i guess i am theoretically using the correct server infomrations. But for some reason the user and password dont fit together, or there is some docker / firewall denying a connection?   https://github.com/supabase/supabase/blob/91eee3106af7a3eebfb8aa3e6a563517ef35285d/docker/.env.example

Running the command with the --local parameter works as expected for generating an empty typescript schema.   

Does anyone have a clue why my connection is getting refused, do i need any other username, kong url or something?

**Comment by theokoel** [comment]
Same issue in my case. I tried to understand how the CLI works. And visibly when you run `supabase gen types typescript`, a docker container is lunch with the [supabase/postgres-meta](https://github.com/supabase/postgres-meta) image.

But when we look this repo, this section is visibly not implement or complet.
![image](https://user-images.githubusercontent.com/41634026/208464335-9d1fc439-f727-4bfc-be2c-45f94c9fc7b5.png)

Maybe `supabase gen types typescript --db-url` is not avalaible for the moment ?
But it looks weird, I must be missing something...
Need to research more about that.

**Reply by Hatzen** [reply]
As far as i understand the Problem, the CLI is not the problem (the readme might not be Up to Date, or the released Version doesnt Match the Open source ones) as i can create the typescript Schema with --local Parameter.  
But the database doesnt seem to be accessible from the postgres-meta container as the Error occures when connecting, before Generating i would say.

**Reply by Hatzen** [reply]
Probably we cannot use localhost as url, because it would refer to the postgres-meta container.
We would need to Pass the postgres-db docker Container IP Adresse (as far as i remember i tried that and it failed as well), probably because the docker instances dont know each other and we would need a Bridge Network, Setup within the docker compose Files.
I will have a Look into that in the Next werk

**Comment by Hatzen** [comment, marked as answer]
Just for documentation purpose, we can use "host.docker.internal" instead of localhost to access the database container from within the postgres-meta container:  
```
npx supabase gen types typescript --schema public --db-url postgresql://postgres:your-super-secret-and-long-postgres-password@host.docker.internal:5432/supabase --debug
``` 
(currently failing on my local setup with user not valid error. but the connection worked so far)  
Another work around is to avoid the supbase cli and just join the docker postgres-meta containers terminal and execute the command to generate the code manually.  

```
node bin/src/server/app.js gen types typescript --include-schemas public
```
